src/misc/unicode.cpp

```diff
@@ -0,0 +1,1159 @@
+};
+
+using code_page_mapping_t = std::map<Grapheme, uint8_t>; // UTF-8 -> code page
+using code_page_mapping_reverse_t = std::map<uint8_t, Grapheme>;
+
+using config_duplicates_t = std::map<uint16_t, uint16_t>;
+using config_aliases_t    = std::vector<std::pair<uint16_t, uint16_t>>;
+
+typedef struct ConfigMappingEntry {
+    bool valid                          = false;
+    code_page_mapping_reverse_t mapping = {};
+    uint16_t extends_code_page          = 0;
+    std::string extends_dir             = "";
+    std::string extends_file            = "";
+} ConfigMappingEntry;
+
+using config_mappings_t = std::map<uint16_t, ConfigMappingEntry>;
+
+static const std::string file_name_main   = "MAIN.TXT";
+static const std::string file_name_ascii  = "ASCII.TXT";
+static const std::string dir_name_mapping = "mapping";
+
+// Thresholds to tell how the code point is encoded
+constexpr uint8_t threshold_non_ascii = 0x80; // 1000 0000
+constexpr uint8_t threshold_encoded   = 0xc0; // 1100 0000
+constexpr uint8_t threshold_bytes_3   = 0xe0; // 1110 0000
+constexpr uint8_t threshold_bytes_4   = 0xf0; // 1111 0000
+constexpr uint8_t threshold_bytes_5   = 0xf8; // 1111 1000
+constexpr uint8_t threshold_bytes_6   = 0xfc; // 1111 1100
+
+// Use the character below if there is absolutely no sane way to handle UTF-8 glyph
+constexpr uint8_t unknown_character = 0x3f; // '?'
+
+// End of file marking, use in some files from unicode.org
+constexpr uint8_t end_of_file_marking = 0x1a;
+
+// Main information about how to create UTF-8 mappings for given DOS code page
+static config_mappings_t config_mappings = {};
+// UTF-8 -> UTF-8 fallback mapping (alias), use before fallback to 7-bit ASCII
+static config_aliases_t config_aliases = {};
+// Information about code pages which are exact duplicates
+static config_duplicates_t config_duplicates = {};
+
+// UTF-8 -> 7-bit ASCII mapping, use as a last resort mapping
+static code_page_mapping_t mapping_ascii = {};
+
+// Concrete UTF-8 -> codepage mappings
+static std::map<uint16_t, code_page_mapping_t> mappings = {};
+// Additional UTF-8 -> codepage mappings, to avoid unknown characters
+static std::map<uint16_t, code_page_mapping_t> mappings_aliases = {};
+// Reverse mappings, codepage -> UTF-8
+static std::map<uint16_t, code_page_mapping_reverse_t> mappings_reverse = {};
+
+// ***************************************************************************
+// Grapheme type implementation
+// ***************************************************************************
+
+static bool is_combining_mark(const uint32_t code_point)
+{
+    static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
+        { 0x0300, 0x036f }, // Combining Diacritical Marks
+        { 0x1ab0, 0x1aff }, // Combining Diacritical Marks Extended
+        { 0x1dc0, 0x1dff }, // Combining Diacritical Marks Supplement
+        { 0x20d0, 0x20ff }, // Combining Diacritical Marks for Symbols
+        { 0xfe20, 0xfe2f }, // Combining Half Marks
```

Comment:
If these are combinations of two 8-bit values (each with notable bit patterns), could also construct them using constexpr bit literals prior, and then build them inside here.

Just an idea - and only if the documentation is describe in bit fields.

Reply:
These are just ranges of code points where combining marks are, without concrete bit patterns - there is no sane way to construct them using bit patterns. See here: https://unicode-table.com/en/blocks/